### PR TITLE
Update HDF5 module

### LINKED ISSUE #11925: hdf5.py fails with exception if the project is not set up for C
**Describe the bug**
what it says in the title. If you have, e.g. a c++ project, trying to call
```dependency('hdf5', language: 'cpp', method: 'config-tool')```
will cause an exception
```
<...>
  File "/usr/lib/python3.11/site-packages/mesonbuild/dependencies/hdf5.py", line 119, in __init__
    compiler = environment.coredata.compilers[for_machine][language]
               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^
KeyError: 'c'
```
hdf5.py always tries to configure ```c``` language, which is a problem whan it's not set in the project

**To Reproduce**
```
project('test', 'cpp')
dependency('hdf5', language: 'cpp', method: 'config-tool')
```

**Expected behavior**
no exception. If C is always needed for HDF5 - make a special error for this case.

**system parameters**
meson 1.0.1
ninja 1.11.1
fedora 38

### PR DESCRIPTION
**Pull request by radistmorse**
Fix wrong env vars and remove redundant C subdependency.

This patch fixes two problems with the HDF5 module. First one is env vars for the HDF5 compile tool. The custom linker is expected to be in `HDF5_CLINKER` var in case of C (see https://github.com/HDFGroup/hdf5/blob/develop/bin/h5cc.in#L97) and in `HDF5_FLINKER` in case of fortran (see https://github.com/HDFGroup/hdf5/blob/develop/fortran/src/h5fc.in#L93).

The second problem is C subdependency for C++ and fortran. The original code for this behaviour was written by @scivision 4 years ago. Unfortunately, he left no contact information, so I wasn't able to ask him why it was implemented the way it was. All my attempts to contact him failed. @scivision if you are reading this, please comment. The way subdependency is implemented now - it is broken (see #11925). Also, it is unneeded. The C libs are indeed needed, which is why both C++ and fortran tools include them in the link list (see https://github.com/HDFGroup/hdf5/blob/develop/c%2B%2B/src/h5c%2B%2B.in#L315 and https://github.com/HDFGroup/hdf5/blob/develop/fortran/src/h5fc.in#L303).

This patch may introduce regressions for the projects that previously set up HDF5 dependency with the wrong `language` parameter, because now you won't get a C-dependency automatically, you need to do it explicitly.

Fixes #11925

### CODE REVIEW

File: mesonbuild/dependencies/hdf5.py [hunk 1]
```diff
@@ -140,13 +143,6 @@ def __init__(self, name: str, environment: 'Environment', kwargs: T.Dict[str, T.
             elif Path(arg).is_file():
                 self.link_args.append(arg)
 
-        # If the language is not C we need to add C as a subdependency
-        if language != 'c':
-            nkwargs = kwargs.copy()
-            nkwargs['language'] = 'c'
-            # I'm being too clever for mypy and pylint
-            self.is_found = self._add_sub_dependency(hdf5_factory(environment, for_machine, nkwargs))  # pylint: disable=no-value-for-parameter
```
Comment (eli-schwartz) [inline mesonbuild/dependencies/hdf5.py:148]: I think this should be its own patch, so 3 in total.

Comment (radistmorse) [reply mesonbuild/dependencies/hdf5.py:148]: This can't be done in this PR, can it? Do I need to open a new one?

Comment (eli-schwartz) [reply mesonbuild/dependencies/hdf5.py:148]: You don't need to push to a new branch and open a new PR. :) You can just prepare the branch locally, and then use `git push -f` instead of `git push` to replace this PR with the new contents.

Comment (radistmorse) [reply mesonbuild/dependencies/hdf5.py:148]: the things you learn about git... Done.

**Comment by eli-schwartz** [comment]
> This patch may introduce regressions for the projects that previously set up HDF5 dependency with the wrong `language` parameter, because now you won't get a C-dependency automatically, you need to do it explicitly.

Unless @scivision clarifies why this is necessary, it seems pretty logical to say that in the general case the C dependency is not needed.

As for cases where you set up the dependency with the wrong language parameter... my assumption is that then you probably do not end up broken anyway, because all that does is add the -lhdf5 which is added either way, so it's merely overlinking. But if you ever didn't end up with the C dependency, you would then fail to build with pkg-config so you should still fix it.

**Review by eli-schwartz — APPROVED** [review]
